### PR TITLE
fix(security): harden agent provisioning execution

### LINKED ISSUE #1707: Security: harden agent provisioning command execution
## Summary

Agent provisioning constructs a shell command from request-derived values before executing the spin-up script. This permits shell interpretation of untrusted input and can expose provisioning secrets through captured script output.

## Fix

- Execute the configured script with an argument array rather than a shell command string.
- Reject unsafe identifiers used in generated endpoint filenames.
- Fail closed when required provisioning configuration is absent.
- Enforce execution timeout and propagate script failures.
- Avoid logging captured script output that may contain secrets.

## Validation

- Added regression tests for argument-array execution, unsafe identifier rejection, and script failure propagation.
- `agent-provisioning` build passes.

### PR DESCRIPTION
**Pull request by sign-mark**
Replace shell command interpolation with `execFile` arguments when provisioning agents. Preserve ledger JSON and local/Docker configuration compatibility; retain identifier validation and safe failure reporting.

Regression tests cover argument handling, ledger configuration and deployment settings, and are included in the CI test selection. The CI-configured tests and both service builds pass locally.

Deploy `agent-service` and `agent-provisioning` together. Live Docker/ECS provisioning still needs retesting. Process-tree cleanup is outside this PR.

Fixes #1707

### CODE REVIEW

**Comment by RinkalBhojani** [comment]
Hey @sign-mark,

Thanks for your contributions. Here are few observations.

You must sign all the commits you are making.  It shows its unverified. Please refer this link for verifying settings at your side - [managing-commit-signature-verification](https://docs.github.com/en/authentication/managing-commit-signature-verification/displaying-verification-statuses-for-all-of-your-commits)
image

Also have a look into coderabbitai review comments and make fixes accordingly wherever applicable.

**Comment by sign-mark** [comment]
@coderabbitai review

**Comment by sign-mark** [comment]
@RinkalBhojani Thanks for your reply, I just signed all the commits, and fixed what coderabbitai reported, please take a look again.

File: apps/agent-provisioning/src/agent-provisioning.service.ts [hunk 1]
```diff
@@ -41,49 +46,79 @@ export class AgentProvisioningService {
...
-        const walletProvision = `${process.cwd() + process.env.AFJ_AGENT_SPIN_UP} ${orgId} "${externalIp}" "${walletName}" "${walletPassword}" ${seed} ${webhookEndpoint} ${walletStorageHost} ${walletStoragePort} ${walletStorageUser} ${walletStoragePassword} ${containerName} ${protocol} ${tenant} ${credoImage} "${indyLedger}" ${inboundEndpoint} ${process.env.SCHEMA_FILE_SERVER_URL} ${process.env.AGENT_API_KEY} ${process.env.AWS_ACCOUNT_ID} ${process.env.S3_BUCKET_ARN} ${process.env.CLUSTER_NAME} ${process.env.TASKDEFINITION_FAMILY} ${process.env.ADMIN_TG_ARN} ${process.env.INBOUND_TG_ARN} ${process.env.FILESYSTEMID} ${process.env.ECS_SUBNET_ID} ${process.env.ECS_SECURITY_GROUP_ID}`;
-        const spinUpResponse: object = new Promise(async (resolve) => {
-          await exec(walletProvision, async (err, stdout, stderr) => {
-            this.logger.log(`shell script output: ${stdout}`);
-            if (stderr) {
-              this.logger.log(`shell script error: ${stderr}`);
-            }
+        this.assertSafeFileIdentifier(orgId, 'orgId');
+        this.assertSafeFileIdentifier(containerName, 'containerName');
```
Comment (ajile-in) [inline apps/agent-provisioning/src/agent-provisioning.service.ts]: [major] This will break provisioning for existing orgs whose names aren't this restrictive. `containerName` comes from `orgData.name` with just spaces swapped for underscores (agent-service.service.ts:420), and org names are free-form — no charset validation anywhere in the chain. So an org named "Acme, Inc.", "O'Reilly Media", "Crédit Agricole" or anything non-Latin provisions fine today but now dies here with "containerName contains unsafe characters".

Since the injection risk is already gone (execFile + argv), maybe normalize instead of reject — NFKD-normalize, replace disallowed chars with `_`, fall back to something safe if the result is empty — and keep this assert as the final guard.

File: apps/agent-provisioning/src/agent-provisioning.service.ts [hunk 2]
```diff
@@ -41,49 +46,79 @@ export class AgentProvisioningService {
...
+            protocol,
+            String(tenant),
+            credoImage,
+            indyLedger,
+            inboundEndpoint,
+            ...requiredEnvironment.map((name) => process.env[name] as string)
+          ],
+          { timeout: AGENT_PROVISION_TIMEOUT_MS, maxBuffer: 1024 * 1024 }
+        ).catch(() => {
+          throw new Error('Agent provisioning script failed');
```
Comment (ajile-in) [inline apps/agent-provisioning/src/agent-provisioning.service.ts]: [minor] This discards everything the script tried to tell us — exit code, signal, all of it — so a prod failure is just "Agent provisioning script failed". It compounds with the outer catch logging `JSON.stringify(error)`: Error objects serialize to `{}` (message/stack are non-enumerable), so the log line ends up `[walletProvision] - error in wallet provision: {}`. Logging `error.message` plus the exit code (still keeping stdout/stderr out, per #1707) would keep failures diagnosable without leaking secrets.

File: apps/agent-provisioning/src/agent-provisioning.service.ts [hunk 3]
```diff
@@ -41,49 +46,79 @@ export class AgentProvisioningService {
...
+            walletStoragePassword,
+            containerName,
+            protocol,
+            String(tenant),
+            credoImage,
+            indyLedger,
+            inboundEndpoint,
+            ...requiredEnvironment.map((name) => process.env[name] as string)
+          ],
+          { timeout: AGENT_PROVISION_TIMEOUT_MS, maxBuffer: 1024 * 1024 }
```
Comment (ajile-in) [inline apps/agent-provisioning/src/agent-provisioning.service.ts]: [note] Heads up that 120s is tighter than the script's own patience — start_agent_ecs.sh polls up to 20×10s (~200s) waiting for the agent to answer, before image pull and ECS registration time on top. A slow start that succeeds today gets killed mid-wait once this lands. Maybe make the timeout env-configurable, or confirm real-world ECS spins fit comfortably under two minutes.

**Review by ajile-in — COMMENTED** [review]
Nice cleanup of a genuinely nasty one — the argv-array switch kills the shell injection from #1707 outright, identifier validation closes the path-traversal angle on the endpoint filename, and the old promise-that-never-rejects hang on script failure is fixed too. Verified the positional args still line up with start_agent.sh ($1–$27), tests pass locally (9/9), typecheck and lint clean.

One regression worth sorting before merge (first comment) — legit org names will now be rejected. Two smaller notes below.

**Comment by sign-mark** [comment]
@ajile-in @RinkalBhojani The latest signed commit `1def944` handles the remaining endpoint-document case by validating the parsed JSON root before reading `CONTROLLER_ENDPOINT`; it also includes a regression test for a literal `null` document. All checks are green. Could you please take a final look?

**Review by ajile-in — APPROVED** [review]
LGTM.

@sign-mark - I have added some commits to fix the pending SonarQube & CodeRabbit issues.

@RinkalBhojani , @ankita-p17 - pls run one manual test and share your comments.

**Comment by ankita-p17** [comment]
> LGTM.
> 
> @sign-mark - I have added some commits to fix the pending SonarQube & CodeRabbit issues.
> 
> @RinkalBhojani , @ankita-p17 - pls run one manual test and share your comments.

We are testing this PR and will update once completed.

**Comment by KambleSahil3** [comment]
@sign-mark While testing this PR I found an issue during agent provisioning. 
For provisioning the agent with indyLedgers,  getting an error while Credo agent configuration file is being generated. The issue appears to be specifically related to the indyLedgers block in the configuration file. Could you please check.

**Comment by sign-mark** [comment]
@KambleSahil3 Fixed the ledger JSON escaping. Please retest with `agent-service` and `agent-provisioning` from this PR together.

I’ve narrowed the changes to the shell-execution fix and its compatibility requirements. Tests and both service builds pass locally; I haven’t tested your Docker/ECS environment.

**Comment by sign-mark** [comment]
Addressed SonarCloud issue `AaA8--C19bZQfg4Ncz5r` (`typescript:S8786`) in a409d666d0e56528b529aad22c0ed29a27743c1f.

Replaced the trailing-underscore regex with a backward index scan. This preserves normalization behavior, including trimming before the 128-character limit, without the repeated regex backtracking on long internal underscore runs.

Validation: all 28 focused tests pass, including 10 new naming boundary cases with 100,000 underscores; 20,000 deterministic inputs produced identical old/new names; ESLint, the agent-provisioning build, signed-commit hooks, and independent review passed. The original regex showed roughly quadratic growth in a local reproduction (8k/16k/32k underscores: approximately 31/116/462 ms).

**Comment by ankita-p17** [comment]
> @KambleSahil3 Fixed the ledger JSON escaping. Please retest with `agent-service` and `agent-provisioning` from this PR together.
> 
> I’ve narrowed the changes to the shell-execution fix and its compatibility requirements. Tests and both service builds pass locally; I haven’t tested your Docker/ECS environment.

Thanks @sign-mark, will test and update
